src/rocket-pool/ObolRocketPoolRecipient.sol

```diff
@@ -0,0 +1,325 @@
+  /// errors
+  /// -----------------------------------------------------------------------
+  /// Invalid minipool deposit amount
+  error InvalidDepositAmount();
+
+  /// Invalid token recovery
+  error InvalidTokenRecovery();
+
+  /// Invalid token recovery recipient
+  error InvalidTokenRecovery_InvalidRecipient();
+
+  /// Invalid minipool
+  error InvalidMinipool_Address();
+
+  /// Invalid distribution
+  error InvalidDistribution_TooLarge();
+
+  /// -----------------------------------------------------------------------
+  /// events
+  /// -----------------------------------------------------------------------
+  /// Emitted after non-ETH tokens are recovered to a recipient
+  /// @param recoveryAddressToken Recovered token
+  /// @param recipient Address receiving recovered token
+  /// @param amount Amount of recovered token
+  event RecoverFunds(address recoveryAddressToken, address recipient, uint256 amount);
+
+  /// Emitted after funds are distributed to recipients
+  /// @param principalPayout Amount of principal paid out
+  /// @param rewardPayout Amount of reward paid out
+  /// @param pullFlowFlag Flag for pushing funds to recipients or storing for
+  /// pulling
+  event DistributeFunds(uint256 principalPayout, uint256 rewardPayout, uint256 pullFlowFlag);
+
+  /// Emitted after funds withdrawn using pull flow
+  /// @param account Account withdrawing funds for
+  /// @param amount Amount withdrawn
+  event Withdrawal(address account, uint256 amount);
+
+  /// -----------------------------------------------------------------------
+  /// storage
+  /// -----------------------------------------------------------------------
+
+  /// -----------------------------------------------------------------------
+  /// storage - constants
+  /// -----------------------------------------------------------------------
+
+  address internal constant ETH_ADDRESS = address(0);
+
+  uint256 internal constant PUSH = 0;
+  uint256 internal constant PULL = 1;
+
+  uint256 internal constant ONE_WORD = 32;
+  uint256 internal constant ADDRESS_BITS = 160;
+
+  /// @dev threshold for pushing balance update as reward or principal
+  uint256 internal constant BALANCE_CLASSIFICATION_THRESHOLD = 4 ether;
+  uint256 internal constant PRINCIPAL_RECIPIENT_INDEX = 0;
+  uint256 internal constant REWARD_RECIPIENT_INDEX = 1;
+
+  /// -----------------------------------------------------------------------
+  /// storage - cwia offsets
+  /// -----------------------------------------------------------------------
+
+  uint256 internal constant STORAGE_ADDRESS_OFFSET = 0;
+
+  // token (address, 20 bytes), recoveryAddress (address, 20 bytes),
+  // tranches (uint256[], numTranches * 32 bytes)
+
+  uint256 internal constant RECOVERY_ADDRESS_OFFSET = 20;
+  // 20 = recoveryAddress_offset (0) + recoveryAddress_size (address, 20
+  // bytes)
+  uint256 internal constant TRANCHES_OFFSET = 40;
+
+  function rpStorage() public pure returns (address) {
```

Comment:
let's store the rpStorage as an immutable variable as it won't change. Using it as a CWIA means we pay gas cost on every txn